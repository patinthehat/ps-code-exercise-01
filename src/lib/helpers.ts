import { StringStatistics } from '@/types/StringStatistics';
import { readFileSync } from 'fs';

export const readTextFile = (path: string): string => readFileSync(path, { encoding: 'utf8' });

export const readTextFileLines = (path: string): string[] =>
    readTextFile(path)
        .trim()
        .split(/\n/)
        .filter(line => line.trim().length > 0);

export const calculateStringStatistics = (str: string): StringStatistics => {
    const alphabet = 'abcdefghijklmnopqrstuvwxyz';
    const vowels = 'aeiou'.split('');
    const consonants = alphabet.split('').filter(letter => !vowels.includes(letter));

    const result = {
        consonants: str.split('').filter(letter => consonants.includes(letter)).length,
        length: str.length,
        vowels: str.split('').filter(letter => vowels.includes(letter)).length,
    };

    return result;
};

export const normalizeStreetAddress = (address: string): string => {
    return address
        .replace(/ Ave\./i, ' Avenue')
        .replace(/ Blvd\./i, ' Boulevard')
        .replace(/ Cir\./i, ' Circle')
        .replace(/ Ct\./i, ' Court')
        .replace(/ Dr\./i, ' Drive')
        .replace(/ Hwy\./i, ' Highway')
        .replace(/ Ln\./i, ' Lane')
        .replace(/ Pkwy\./i, ' Parkway')
        .replace(/ Pl\./i, ' Place')
        .replace(/ Rd\./i, ' Road')
        .replace(/ St\./i, ' Street');
};
