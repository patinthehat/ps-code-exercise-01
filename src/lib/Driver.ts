import { calculateStringStatistics } from '@/lib/helpers';
import { Shipment } from '@/lib/Shipment';

export class Driver {
    public name: string;
    public shipment: Shipment | null = null;
    public numberOfVowelsInName = 0;
    public numberOfConsonantsInName = 0;

    constructor(name: string) {
        this.name = name;
        this.calculateVowelsAndConsonantsInName();
    }

    public calculateVowelsAndConsonantsInName(): void {
        const data = calculateStringStatistics(this.name);

        this.numberOfVowelsInName = data.vowels;
        this.numberOfConsonantsInName = data.consonants;
    }

    public suitabilityScoreForShipment(shipment: Shipment): number {
        return shipment.suitabilityScoreForDriver(this);
    }
}
