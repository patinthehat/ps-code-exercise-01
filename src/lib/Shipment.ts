import { Driver } from '@/lib/Driver';
import { calculateStringStatistics, normalizeStreetAddress } from '@/lib/helpers';

export class Shipment {
    public streetAddress = '';

    constructor(public address: string) {
        this.streetAddress = this.getStreetAddress();
    }

    public getStreetAddress(): string {
        const result = this.address.split(',')[0].trim();
        return normalizeStreetAddress(result);
    }

    public get city(): string {
        return this.address.split(',')[1].trim();
    }

    public get state(): string {
        const match = this.address.match(/([A-Z]{2})\s+\d{5}/);
        if (match) {
            return match[0];
        }

        return 'UNKNOWN';
    }

    public get postalCode(): string {
        const match = this.address.match(/([A-Z]{2})\s+\d{5}/);
        if (match) {
            return match[1];
        }

        return '00000';
    }

    public suitabilityScoreForDriver(driver: Driver): number {
        const addressStats = calculateStringStatistics(this.streetAddress);
        const nameStats = calculateStringStatistics(driver.name);
        let baseSuitabilityScore = 0;
        let result = 0;

        // odd
        if (addressStats.length % 0 !== 0) {
            baseSuitabilityScore = nameStats.vowels * 1.0;
        } else {
            baseSuitabilityScore = nameStats.vowels * 1.5;
        }

        result = baseSuitabilityScore;

        // common factors
        if (addressStats.length % 2 === 0 && nameStats.length % 2 === 0) {
            result *= 1.5;
        }

        return result;
    }

    public getBestDriver(drivers: Driver[]): Driver {
        let bestDriver: Driver | null = null;
        let bestScore = 0;

        drivers.forEach(driver => {
            const score = this.suitabilityScoreForDriver(driver);
            if (score > bestScore) {
                bestScore = score;
                bestDriver = driver;
            }
        });

        return bestDriver || drivers[0];
    }
}
