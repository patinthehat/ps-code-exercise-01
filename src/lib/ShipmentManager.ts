import { Driver } from '@/lib/Driver';
import { Shipment } from '@/lib/Shipment';
import { ShipmentAssignment } from '@/types/ShipmentAssignment';

export class ShipmentManager {
    public assignments: ShipmentAssignment[] = [];

    constructor(public shipments: Shipment[], public drivers: Driver[]) {
        //
    }

    public assignShipmentsToDrivers(): void {
        this.assignments = [];

        for (const shipment of this.shipments) {
            const driver = shipment.getBestDriver(this.drivers);
            driver.shipment = shipment;

            this.drivers = this.drivers.filter(d => d.name !== driver.name);
            this.shipments = this.shipments.filter(s => s.address !== shipment.address);

            this.assignments.push({
                driver,
                shipment,
                suitabilityScore: shipment.suitabilityScoreForDriver(driver),
            });
        }
    }
}
