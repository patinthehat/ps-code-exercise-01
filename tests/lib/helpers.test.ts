import { calculateStringStatistics, normalizeStreetAddress, readTextFileLines } from '@/lib/helpers';

it('calculates the correct number of vowels in a string', () => {
    expect(calculateStringStatistics('hello').vowels).toBe(2);
});

it('calculates the correct number of consonants in a string', () => {
    expect(calculateStringStatistics('hello').consonants).toBe(3);
});

it('calculates the correct number of vowels + consonants in a string', () => {
    const stats = calculateStringStatistics('hello');
    expect(stats.consonants + stats.vowels).toBe(5);
});

it('calculates the correct length of a string', () => {
    expect(calculateStringStatistics('hello').length).toBe(5);
});

it('reads the correct number of lines from a text file', () => {
    expect(readTextFileLines(`${__dirname}/../fixtures/drivers-01.txt`).length).toBe(15);
});

it('reads the correct contents from a text file', () => {
    expect(readTextFileLines(`${__dirname}/../fixtures/drivers-01.txt`)).toMatchSnapshot();
});

it('normalizes street addresses', () => {
    const addresses = [
        '123 Main St.',
        '123 Main Ave.',
        '1231 Yonge Street',
        '456 West 1st Rd.',
        '9371 Eastwood Avenue'
    ];

    const expectedAddress = [
        '123 Main Street',
        '123 Main Avenue',
        '1231 Yonge Street',
        '456 West 1st Road',
        '9371 Eastwood Avenue'
    ];

    addresses.forEach((address, index) => {
        expect(normalizeStreetAddress(address)).toBe(expectedAddress[index]);
    });
});
