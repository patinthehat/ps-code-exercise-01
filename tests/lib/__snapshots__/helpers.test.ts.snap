// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reads the correct contents from a text file 1`] = `
[
  "Malcolm Parker",
  "Zachariah Hardwick",
  "Cory Paxton",
  "Yancy Bloodworth",
  "Star Baldwin",
  "Elwood Bryson",
  "Katheryn Bancroft",
  "Juliana Brassington",
  "Isadore Thompson",
  "Jade Ashley",
  "Noelle Evelyn",
  "Darcie Gary",
  "Sherwood Barret",
  "Winifred Boatwright",
  "Deanna Bell",
]
`;
